### PR TITLE
fix: clean up feature shortcodes in .html files and resolve orphaned …

### LINKED ISSUE #4684: `make feature-shortcode-update` skips non-`.md` files and orphans shortcodes added after their version ships
**Problem 1: only `.md` files are processed**

`build/scripts/feature-shortcode-update/main.go:45` skips everything else:

```go
if d.IsDir() || !strings.HasSuffix(d.Name(), ".md") {
    return nil
}
```

`site/content/en/docs/Reference/agones_crd_api_reference.html` lives in the same tree and uses the same shortcodes, so the release-time cleanup never reaches it.

The two-block structure that `gen-api-docs.sh` maintains is correct during a dev cycle — it re-tags the old publish block as `expiryVersion=$VERSION` and appends newly-generated content as `publishVersion=$VERSION`, which is what lets the released site and the dev site render different API docs from one source. But at release the `expiryVersion` half becomes permanently hidden and should be dropped, the same way it is for `.md`. It hasn't been:

| Tag | Shortcodes | Lines |
|---|---|---|
| v1.57.0 | `expiryVersion="1.57.0"` / `publishVersion="1.57.0"` | 10027 |
| v1.58.0 | `expiryVersion="1.57.0"` / `publishVersion="1.57.0"` | 10027 |
| v1.59.0 | `expiryVersion="1.59.0"` / `publishVersion="1.59.0"` | 9993 |

Each of those tags ships ~5k lines that render at no version. In #4683 this had to be removed by hand.

Caveat for implementation: `gen-api-docs.sh` locates its baseline by scanning for the `publishVersion` shortcode:

```bash
awk '/ feature publishVersion/{flag=1;next}/ \/feature/{flag=0}flag' $FILE > $OLD
```

so the *publish* wrapper must survive the cleanup, or the generator will rewrite the file and `test-gen-api-docs` will fail with a misleading "API docs are out of date". Only the expiry block should be removed.

**Problem 2: exact version matching orphans late-added shortcodes**

The script only removes shortcodes matching the exact `-version` passed, so a shortcode naming an already-shipped version never matches a future release. Two live examples on `main`:

| File | Shortcode | Added | Version released |
|---|---|---|---|
| `site/content/en/docs/Guides/Client SDKs/rust.md` | `1.52.0` pair | 2025-09-10 (#4247) | 2025-09-09 |
| `site/content/en/docs/Reference/fleetautoscaler.md` | `1.54.0` pair | 2025-12-13 (#4365) | 2025-12-02 |

Both landed after their release had already run the cleanup. They render correctly but are permanent dead markup.

**Suggested direction**

- Widen the walk to `.html`, removing only the `expiryVersion` block (see caveat above).
- Resolve every shortcode with a version `<=` the release version, not just exact matches.
- Optionally, a CI check flagging shortcodes that name an already-released version.

**Constraint**

`site/content/en/docs/Contribute/documentation-editing-contribution.md` contains *escaped* examples (`{{%/* feature publishVersion="1.24.0" */%}}`) documenting the shortcode itself. These must not be rewritten.

**Files**

- `build/scripts/feature-shortcode-update/main.go`
- `build/includes/website.mk` (`feature-shortcode-update`, lines 118-120)
- `docs/governance/templates/release_issue.md`

### PR DESCRIPTION
**Pull request by Sivasankaran25**
…versions

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

 /kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #4684

**Did you use AI tools in preparing this PR?**:

Y/N

**Special notes for your reviewer**:

### CODE REVIEW

File: build/scripts/feature-shortcode-update/main.go [hunk 1]
```diff
@@ -98,61 +115,101 @@ func main() {
...
+	toInt := func(s string) (int, error) {
+		return strconv.Atoi(strings.TrimSpace(s))
+	}
+
+	for i := 0; i < len(vParts) || i < len(tParts); i++ {
+		var vn, tn int
+		var err error
+		if i < len(vParts) {
+			if vn, err = toInt(vParts[i]); err != nil {
+				log.Printf("warning: could not parse version segment %q in %q, treating as 0", vParts[i], v)
```
Comment (igooch) [inline build/scripts/feature-shortcode-update/main.go]: A version segment that fails to parse is logged and then treated as `0`, which makes the block look *older* than the release and deletes it.

Reproduced on this branch:

```
$ cat t.md
intro
{{% feature expiryVersion="1.x" %}}
FUTURE CONTENT DO NOT DELETE
{{% /feature %}}
outro

$ go run main.go -version=1.61.0
warning: could not parse version segment "x" in "1.x", treating as 0
Processed file: site/content/en/docs/t.md

$ cat t.md
intro
outro
```

`expiryVersion="v1.61.0"` does the same thing — `Atoi("v1")` fails, `0 < 1`, block removed. So a single-character typo in a shortcode silently drops documentation during a release cut, and the only signal is a `log.Printf` line in the middle of the tool's output.

Safer default is to refuse to resolve what you can't parse:

```go
func versionLTE(v, target string) (bool, error) {
	// ... return an error on the first segment that fails Atoi
}
```

and at the call sites, `log.Printf("skipping %s: unparseable version %q in %s", ...)` and keep the block. Leaving it in place is recoverable (a human sees the stale shortcode next release); deleting it is not.

File: docs/governance/templates/release_issue.md [hunk 2]
```diff
@@ -37,7 +37,7 @@ and copy it into a release issue. Fill in relevant values, found inside {}
...
   - [ ] Run `make del-data-proofer-ignore FILENAME={version}-1.md` to remove `data-proofer-ignore` attribute from previous release blog. Review all occurrences of the link_test and data-proofer-ignore attributes globally. Exclude html and release files.
-  - [ ] Run `make feature-shortcode-update version={version}` to remove all instances of the `feature expiryVersion` shortcode, including the associated content, while preserving the rest of the content within the .md files located in site/content/en/docs. Additionally, ensure that only the block of `feature publishVersion` is removed without affecting the content.
+  - [ ] Run `make feature-shortcode-update version={version}` to remove all `feature expiryVersion` shortcodes (tags and content) and unwrap all `feature publishVersion` shortcodes (tags only, content kept) whose version is <= {version}, across both .md and .html files in site/content/en/docs.
```
Comment (igooch) [inline docs/governance/templates/release_issue.md]: Same issue as the `website.mk` comment, and it matters more here because this is the checklist someone follows under release pressure. "unwrap all `feature publishVersion` shortcodes (tags only, content kept) ... across both .md and .html files" is not what the script does — the publish wrapper in `agones_crd_api_reference.html` is deliberately left in place.

If the release engineer reads this and then sees `{{% feature publishVersion="..." %}}` still sitting at the top of the CRD reference, the natural reaction is to delete it by hand — which breaks `test-gen-api-docs` with the misleading "API docs are out of date" error described in #4684.

Suggested wording:

> Run `make feature-shortcode-update version={version}` to resolve every `feature` shortcode naming a version `<= {version}` in `site/content/en/docs`: `expiryVersion` blocks are removed along with their content (both `.md` and `.html`), and `publishVersion` blocks are unwrapped, keeping their content (`.md` only). The `publishVersion` wrapper in `agones_crd_api_reference.html` is intentionally left in place — `gen-api-docs.sh` uses it to locate the previous version's API docs, so removing it will break `make test-gen-api-docs`.

File: build/scripts/feature-shortcode-update/main.go [hunk 3]
```diff
@@ -98,61 +115,101 @@ func main() {
 	}
 }
 
-func removeBlocks(scanner *bufio.Scanner, version string) string {
+// removeBlocks assumes feature shortcodes never nest — a {{%|{{< /feature %}}|>}} always
+// closes the single innermost open block, tracked via inExpiryBlock/inPublishBlock below.
+func removeBlocks(scanner *bufio.Scanner, targetVersion, ext string) string {
```
Comment (igooch) [inline build/scripts/feature-shortcode-update/main.go]: This script has no `main_test.go`, recommend writing a new test as this PR adds a couple edge cases. There's a number of `main_test.go` in `build/scripts`, so there's a pattern to follow.

Cases worth pinning down, most of which came out of the Copilot threads on this PR:

- `versionLTE`: `1.9.0` vs `1.10.0` (the numeric-vs-string case this PR fixes), equal versions, differing segment counts (`1.61` vs `1.61.0`), and whatever you decide unparseable input should do.
- `removeBlocks` on `.md`: resolved expiry removed with content; resolved publish unwrapped with content kept; a future-version block of each kind left completely alone.
- `removeBlocks` on `.html`: expiry removed, publish wrapper preserved — this is the one that keeps `test-gen-api-docs` green, so it's worth locking down explicitly.
- The escaped `{{%/* feature ... */%}}` form is not rewritten.
- Both `{{% %}}` and `{{< >}}` delimiter styles.

File: build/scripts/feature-shortcode-update/main.go [hunk 4]
```diff
@@ -98,61 +110,96 @@ func main() {
...
-			publishBlock = false
-			modified = true
-			continue
+		if ext == ".md" {
+			if v, ok := matchOpen(line, publishOpenPercentRe, publishOpenAngleRe); ok && versionLTE(v, targetVersion) {
+				inPublishBlock = true
+				modified = true
+				continue
+			}
 		}
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go]: Agreed there's a mismatch, but I'd resolve it the other way round — **fix the docs, keep the `.md`-only restriction**. Removing it would break the release.

`site/gen-api-docs.sh:48` locates the previous version's API docs by scanning for the publish shortcode:

```bash
awk '/\ feature\ publishVersion/{flag=1;next}/\ \/feature/{flag=0}flag' $FILE > $OLD
```

If the cleanup unwrapped publishVersion in `agones_crd_api_reference.html`, that `awk` would produce an empty `$OLD`, the generator would rewrite the file, and `make test-gen-api-docs` would fail with the misleading "API docs are out of date" error. #4684 calls this out explicitly. I confirmed the current behaviour is correct with a synthetic fixture — expiry half removed, publish wrapper preserved:

```
+++
title="Agones Kubernetes API"
+++

{{% feature publishVersion="1.61.0" %}}
<p>NEW DOCS</p>
{{% /feature %}}
```

So the two doc strings are what need correcting. Suggested wording —

`build/includes/website.mk`:

```make
# For resolved shortcodes (version <= the release being cut) in site/content/en/docs:
#   - expiryVersion: remove the tags and the wrapped content (.md and .html)
#   - publishVersion: remove the tags, keep the content (.md only — .html keeps its
#     publish wrapper, which gen-api-docs.sh uses to locate the previous API docs)
```

`docs/governance/templates/release_issue.md:40` — this one matters more, because it's the checklist someone follows under release pressure. If it says publishVersion is unwrapped in `.html` and they then see `{{% feature publishVersion="..." %}}` still sitting at the top of the CRD reference, the natural reaction is to delete it by hand and break `test-gen-api-docs`:

> Run `make feature-shortcode-update version={version}` to resolve every `feature` shortcode naming a version `<= {version}` in `site/content/en/docs`: `expiryVersion` blocks are removed along with their content (both `.md` and `.html`), and `publishVersion` blocks are unwrapped, keeping their content (`.md` only). The `publishVersion` wrapper in `agones_crd_api_reference.html` is intentionally left in place — `gen-api-docs.sh` uses it to locate the previous version's API docs, so removing it will break `make test-gen-api-docs`.

One related thought on the code: the exemption is keyed on the *extension* rather than on the one file that needs it. `agones_crd_api_reference.html` is the only `.html` under `site/content/en/docs` today, so a second one would get its publishVersion silently orphaned forever — the exact failure mode this PR exists to fix, just relocated. Either scope it to the filename, or keep the extension gate and `log.Printf` a warning when a resolved publishVersion is skipped in an `.html` file, so it surfaces in the release output.

File: build/scripts/feature-shortcode-update/main.go [hunk 5]
```diff
@@ -98,61 +110,96 @@ func main() {
...
-			expiryBlock = true
-			preserveLines = false
+		if inExpiryBlock {
+			// Drop every line inside a resolved expiryVersion block, including its own closing tag.
+			if matchFeatureClose(line) {
+				inExpiryBlock = false
+			}
 			modified = true
 			continue
 		}
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go:130]: Confirmed, and the outcome is worse than "gets stuck" — it destroys content that was never inside a block. Ran this on the branch:

```
$ cat site/content/en/docs/oneline.md
a
{{% feature expiryVersion="1.50.0" %}}x{{% /feature %}}
b
c

$ go run main.go -version=1.61.0
Processed file: site/content/en/docs/oneline.md

$ cat site/content/en/docs/oneline.md
a
```

`b` and `c` are gone. The open tag sets `inExpiryBlock`, the close on that same line is never examined, and every subsequent line is dropped to EOF. In a file that *does* have a later block, it instead swallows that block's closing tag and leaves stray markup.

One correction to the framing though: **this is not a regression from this PR.** I ran the identical fixture against `main` and got the same single line of output — the old `preserveLines` flag has exactly the same hole. So it's pre-existing, and there's no single-line usage anywhere in `site/content/en/docs` today (all five occurrences are on their own lines), which is why it has never bitten.

That makes it a "should fix" rather than a blocker for this PR, and @Sivasankaran25 it's reasonable to split it into a follow-up if you'd rather keep this change scoped. If you do fix it here, the regex refactor makes it cheap — check `matchFeatureClose(line)` on the opening line before setting the flag:

```go
if v, ok := matchOpen(line, expiryOpenPercentRe, expiryOpenAngleRe); ok && versionLTE(v, targetVersion) {
	inExpiryBlock = !matchFeatureClose(line)
	modified = true
	continue
}
```

Note the publish case needs different handling, not the same line of code — see the sibling thread.

File: build/scripts/feature-shortcode-update/main.go [hunk 6]
```diff
@@ -98,61 +110,96 @@ func main() {
...
 
-		// Check if the line contains the starting of the publishVersion shortcode with the specified version
-		if strings.Contains(line, "{{% feature publishVersion=\""+version+"\" %}}") ||
-			strings.Contains(line, "{{< feature publishVersion=\""+version+"\" >}}") {
-			publishBlock = true
+		if v, ok := matchOpen(line, expiryOpenPercentRe, expiryOpenAngleRe); ok && versionLTE(v, targetVersion) {
+			inExpiryBlock = true
 			modified = true
 			continue
 		}
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go:154]: Same finding as the thread on the expiry branch above — I've put the reproduction and the "this also affects `main`, so it's pre-existing" evidence there to keep it in one place.

Worth flagging that the publish case does **not** fail the same way, so "apply the same reasoning" needs a different fix. For a single-line `{{% feature publishVersion="1.5.0" %}}text{{% /feature %}}`, the whole line is dropped by the `continue`, so the wrapped `text` is lost even though publish blocks are supposed to keep their content — and then `inPublishBlock` stays set and consumes the next unrelated closing tag. Unwrapping in place is what's actually wanted here, i.e. strip the tags from the line and keep the remainder, rather than just clearing the flag.

File: build/scripts/feature-shortcode-update/main.go [hunk 7]
```diff
@@ -98,61 +110,96 @@ func main() {
...
+		return n
+	}
+
+	for i := 0; i < len(vParts) || i < len(tParts); i++ {
+		var vn, tn int
+		if i < len(vParts) {
+			vn = toInt(vParts[i])
+		}
+		if i < len(tParts) {
+			tn = toInt(tParts[i])
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go]: Agreed, and I'd call this the one blocking item on the PR. It isn't only a misclassification — coercing to `0` makes an unparseable version look *older* than the release, so the block and its content get deleted:

```
$ cat site/content/en/docs/t.md
intro
{{% feature expiryVersion="1.x" %}}
FUTURE CONTENT DO NOT DELETE
{{% /feature %}}
outro

$ go run main.go -version=1.61.0
warning: could not parse version segment "x" in "1.x", treating as 0
Processed file: site/content/en/docs/t.md

$ cat site/content/en/docs/t.md
intro
outro
```

The realistic trigger isn't `-rc1` suffixes, it's a plain typo. `expiryVersion="v1.61.0"` does the same thing: `Atoi("v1")` fails, first segment compares `0 < 1`, block removed. So one stray character in a shortcode silently drops documentation during a release cut, and the only signal is a `log.Printf` line buried in the output of a tool that walks the whole docs tree.

On "failing fast" — I'd lean toward **failing closed rather than failing fast**: return an error from `versionLTE`, and at the call site log which file and shortcode were skipped and leave the block in place.

```go
func versionLTE(v, target string) (bool, error) {
	// return an error on the first segment that fails Atoi
}
```

Leaving a stale shortcode in place is recoverable — someone notices it next release, which is exactly the class of thing this PR is fixing. Deleting content is not, and `make feature-shortcode-update` is run mid-release when the diff is competing for attention with a dozen other checklist items. Hard-failing the whole run is also defensible, but it means one typo blocks the release cut on a cleanup step.

File: build/scripts/feature-shortcode-update/main.go [hunk 8]
```diff
@@ -57,7 +69,7 @@ func main() {
 		}()
 
 		scanner := bufio.NewScanner(file)
-		modifiedContent := removeBlocks(scanner, *version)
+		modifiedContent := removeBlocks(scanner, *version, ext)
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go]: Agreed on both halves, with a note on severity.

Not currently reachable — the longest line anywhere under `site/content/en/docs` is 1358 chars (in `Getting Started/create-webhook-fleetautoscaler.md`), and `agones_crd_api_reference.html` is generated by `gen-crd-api-reference-docs` with one tag per line. So this is a latent issue rather than a live break.

The reason it's still worth fixing here is that the failure isn't just "partial processing" — the partial result gets **written back over the file**. If the scan aborts after `modified` has already been set by an earlier resolved block, `main` takes the truncated string and `os.Create`s the file with it. Silent data loss with a zero exit code.

`scanner.Buffer(...)` plus `scanner.Err()` covers it; the important part is that the error reaches `main` and short-circuits before the `os.Create` on line 81, which means `removeBlocks` returning `(string, error)`.

File: build/scripts/feature-shortcode-update/main.go [hunk 9]
```diff
@@ -57,7 +69,7 @@ func main() {
...
 
 		// Only write the modified content back to the .md file if there are changes
 		if modifiedContent != "" {
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go]: Confirmed reachable. A file whose entire contents are a resolved block comes out unchanged:

```
$ cat site/content/en/docs/only.md
{{% feature expiryVersion="1.50.0" %}}
all of it
{{% /feature %}}

$ go run main.go -version=1.61.0
There are no files with feature expiryVersion or publishVersion shortcodes

$ cat site/content/en/docs/only.md
{{% feature expiryVersion="1.50.0" %}}
all of it
{{% /feature %}}
```

The shortcode survives the cleanup silently, and the log line actively says the opposite of what happened.

In practice every `.md` under `site/content/en/docs` has front matter, so a fully-empty result is unlikely today. But the sentinel-value conflation is worth removing regardless, and `removeBlocks` is likely to grow an `error` return anyway for the `scanner.Err()` thread — so `(string, bool, error)` or a small result struct lands both fixes together.

File: build/scripts/feature-shortcode-update/main.go [hunk 10]
```diff
@@ -18,141 +18,227 @@ package main
...
+	toInt := func(s string) (int, error) {
+		return strconv.Atoi(strings.TrimSpace(s))
+	}
+
+	for i := 0; i < len(vParts) || i < len(tParts); i++ {
+		var vn, tn int
+		var err error
+		if i < len(vParts) {
+			if vn, err = toInt(vParts[i]); err != nil {
+				log.Printf("warning: could not parse version segment %q in %q, treating as 0", vParts[i], v)
```
Comment (igooch) [inline build/scripts/feature-shortcode-update/main.go]: This is the blocking item from my previous review, still unresolved: an unparseable segment is coerced to 0, which makes the version compare as *older* than the release, so the caller deletes the expiry block's content (or unwraps a publish block) on a typo. The reproduction from last round still holds — `expiryVersion="1.x"` gets its content deleted with only a log warning scrolling past in release output.

A release cleanup must fail closed on input it can't parse. Concretely: have `versionLTE` return `(bool, error)` (or validate all shortcode versions up front) and abort the run — `main` already propagates errors from the walk, so this now fits the structure cheaply. Skipping the block with a warning would also be acceptable (it just leaves dead markup, which the next release sweeps up); silently deleting content is the one behavior that shouldn't survive.

File: build/scripts/feature-shortcode-update/main_test.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+		target   string
+		expected bool
+	}{
+		{"numeric vs string ordering", "1.9.0", "1.10.0", true}, // the bug this PR fixes
+		{"equal versions", "1.61.0", "1.61.0", true},
+		{"differing segment counts, v shorter", "1.61", "1.61.0", true},
+		{"differing segment counts, v longer but equal prefix", "1.61.0", "1.61", true},
+		{"v greater than target", "1.62.0", "1.61.0", false},
+		{"v less than target", "1.5.0", "1.61.0", true},
+		{"unparseable segment in v treated as 0", "1.x.0", "1.1.0", true},
```
Comment (igooch) [inline build/scripts/feature-shortcode-update/main_test.go]: These two cases pin the coercion-to-0 behavior as *intended* — but per the thread on `versionLTE`, that behavior deletes content on a typo'd version and needs to become an error (or a skip). When that changes, these cases should assert the failure mode instead: unparseable input in `v` or `target` yields an error / leaves the block untouched, rather than `true`.

File: build/scripts/feature-shortcode-update/main.go [hunk 12]
```diff
@@ -18,141 +18,227 @@ package main
...
-		// Check if the line contains the ending of the expiryVersion shortcode
-		if (strings.Contains(line, "{{% /feature %}}") && expiryBlock) ||
-			(strings.Contains(line, "{{< /feature >}}") && expiryBlock) {
-			expiryBlock = false
-			preserveLines = true
-			modified = true
+		if inPublishBlock {
+			// Only reached for .md files (see below) - drop just the
+			// closing tag, content is kept.
+			if matchFeatureClose(line) {
```
Comment (igooch) [inline build/scripts/feature-shortcode-update/main.go:135]: New content-loss case introduced by this rework: when the *closing* tag shares a line with content, the whole line is dropped, taking the content with it.

```
{{% feature publishVersion="1.50.0" %}}
keep this{{% /feature %}}
```

`keep this` is deleted, but publish blocks are supposed to keep their content. The open-tag side already handles this via `stripPublishTagsOnLine`; the close side should do the equivalent — strip just the close tag and emit the remainder:

```go
if matchFeatureClose(line) {
	inPublishBlock = false
	modified = true
	rest := featureClosePercentRe.ReplaceAllString(line, "")
	rest = featureCloseAngleRe.ReplaceAllString(rest, "")
	if strings.TrimSpace(rest) != "" {
		sb.WriteString(rest)
		sb.WriteString("\n")
	}
	continue
}
```

Worth a test case alongside the existing single-line ones.

File: build/scripts/feature-shortcode-update/main.go [hunk 13]
```diff
@@ -18,141 +18,227 @@ package main
...
 			continue
 		}
 
-		// Check if the line contains the starting of the publishVersion shortcode with the specified version
-		if strings.Contains(line, "{{% feature publishVersion=\""+version+"\" %}}") ||
-			strings.Contains(line, "{{< feature publishVersion=\""+version+"\" >}}") {
-			publishBlock = true
+		if v, ok := matchOpen(line, expiryOpenPercentRe, expiryOpenAngleRe); ok && versionLTE(v, targetVersion) {
 			modified = true
+			if matchFeatureClose(line) {
```
Comment (igooch) [inline build/scripts/feature-shortcode-update/main.go]: Nit, non-blocking: this drops the entire line, so `prefix {{% feature expiryVersion="1.50.0" %}}x{{% /feature %}} suffix` loses `prefix`/`suffix` (Copilot flagged this too). No current doc content mixes prose and shortcodes on one line, and the old code had the same line-granularity behavior, so I'm fine deferring — but if you fix the publish close-tag case above with span-stripping, doing the same here makes the two paths consistent.

File: docs/governance/templates/release_issue.md (hunk 2 above)
Comment (igooch) [inline docs/governance/templates/release_issue.md]: This says publish shortcodes are unwrapped "across both .md and .html", but the implementation deliberately unwraps in `.md` only — the `.html` publish wrapper is the marker `site/gen-api-docs.sh` scans for. As written, a release operator checking the output against this step will think the tool missed `agones_crd_api_reference.html` and be tempted to unwrap it by hand, which breaks `test-gen-api-docs`.

Suggested wording: "…remove all `feature expiryVersion` shortcodes (tags and content) whose version is <= {version} from .md and .html files, and unwrap `feature publishVersion` shortcodes (tags only, content kept) in .md files only — publishVersion wrappers in .html are intentionally preserved because `site/gen-api-docs.sh` uses them to locate the current API docs."

File: build/includes/website.mk [hunk 14]
```diff
@@ -113,7 +113,8 @@ test-gen-api-docs: ensure-build-image
 	@diff -bB /tmp/result.sorted /tmp/generated.html.sorted || \
 		(echo "\n\n>>> Error: API docs are out of date. Please run 'make gen-api-docs' to regenerate them since the CRD information in /pkg/apis has changed.\n\n" && exit 1)
 
-# Remove feature expiry/publish version shortcodes update in site/content/en/docs
+# Remove resolved feature expiryVersion/publishVersion shortcodes (version <= the release being cut)
+# from .md and .html files in site/content/en/docs.
```
Comment (igooch) [inline build/includes/website.mk]: Same doc gap as the release template: worth one more line here noting the exception, e.g. "publishVersion wrappers in .html are preserved — gen-api-docs.sh depends on them." Cheap insurance against someone 'simplifying' the ext check later without knowing why it exists.

[the remark this answers is not in the document]
Comment (igooch) [reply build/includes/website.mk]: +1 It's a cheap fix that removes ambiguity.

File: build/scripts/feature-shortcode-update/main.go [hunk 15]
```diff
@@ -18,141 +18,227 @@ package main
...
+				// Whole block resolves on one line: drop it entirely,
+				// stay out of any block state.
+				continue
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go]: +1, although this is a "nit" rather that "high". No current doc content mixes prose and shortcodes on one line, and the old code had the same line-granularity behavior, so I'm fine deferring — but if you fix the publish close-tag case above with span-stripping, doing the same here makes the two paths consistent.

File: build/scripts/feature-shortcode-update/main.go [hunk 16]
```diff
@@ -18,141 +18,227 @@ package main
...
+		var vn, tn int
+		var err error
+		if i < len(vParts) {
+			if vn, err = toInt(vParts[i]); err != nil {
+				log.Printf("warning: could not parse version segment %q in %q, treating as 0", vParts[i], v)
+			}
+		}
+		if i < len(tParts) {
+			if tn, err = toInt(tParts[i]); err != nil {
+				log.Printf("warning: could not parse version segment %q in %q, treating as 0", tParts[i], target)
```
[the remark this answers is not in the document]
Comment (igooch) [reply build/scripts/feature-shortcode-update/main.go]: +1 A release cleanup must fail closed on input it can't parse. Concretely: have versionLTE return (bool, error) (or validate all shortcode versions up front) and abort the run — main already propagates errors from the walk, so this now fits the structure cheaply. Skipping the block with a warning would also be acceptable (it just leaves dead markup, which the next release sweeps up); silently deleting content is the one behavior that shouldn't survive.

Please also update the related test cases.

File: docs/governance/templates/release_issue.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (igooch) [reply docs/governance/templates/release_issue.md]: +1 It's a cheap fix that removes ambiguity.
